Gems/MachineLearning/Code/Tests/Algorithms/ActivationTests.cpp

```diff
@@ -0,0 +1,110 @@
+    {
+    };
+
+    TEST_F(MachineLearning_Activations, OneHotArgMax)
+    {
+        AZStd::size_t testValue = 1;
+        AZ::VectorN testVector;
+
+        MachineLearning::OneHotEncode(testValue, 10, testVector);
+        EXPECT_FLOAT_EQ(testVector.GetElement(0), 0.0f);
+        EXPECT_FLOAT_EQ(testVector.GetElement(1), 1.0f);
+        EXPECT_EQ(MachineLearning::ArgMaxDecode(testVector), testValue);
+
+        testValue = 3;
+        MachineLearning::OneHotEncode(testValue, 10, testVector);
+        EXPECT_EQ(MachineLearning::ArgMaxDecode(testVector), testValue);
+
+        testValue = 7;
+        MachineLearning::OneHotEncode(testValue, 10, testVector);
+        EXPECT_EQ(MachineLearning::ArgMaxDecode(testVector), testValue);
+
+        testValue = 8;
+        MachineLearning::OneHotEncode(testValue, 10, testVector);
+        EXPECT_EQ(MachineLearning::ArgMaxDecode(testVector), testValue);
+    }
+
+    TEST_F(MachineLearning_Activations, TestRelu)
+    {
+        AZ::VectorN output = AZ::VectorN::CreateZero(1024);
+        AZ::VectorN sourceVector = AZ::VectorN::CreateRandom(1024);
+        sourceVector *= 100.0f;
+        sourceVector -= 50.0f;
+        MachineLearning::ReLU(sourceVector, output);
+    
+        for (AZStd::size_t iter = 0; iter < output.GetDimensionality(); ++iter)
+        {
+            ASSERT_GE(output.GetElement(iter), 0.0f);
+        }
+    }
+
+    TEST_F(MachineLearning_Activations, TestSigmoid)
+    {
+        AZ::VectorN output = AZ::VectorN::CreateZero(1024);
+        AZ::VectorN sourceVector = AZ::VectorN::CreateRandom(1024);
+        sourceVector *= 100.0f;
+        sourceVector -= 50.0f;
+        MachineLearning::Sigmoid(sourceVector, output);
+
+        // Sigmoid guarantees all outputs get squished between 0 and 1
+        for (AZStd::size_t iter = 0; iter < output.GetDimensionality(); ++iter)
+        {
+            ASSERT_GE(output.GetElement(iter), 0.0f);
+            ASSERT_LE(output.GetElement(iter), 1.0f);
+        }
+    }
+
+    TEST_F(MachineLearning_Activations, TestSoftmax)
+    {
+        AZ::VectorN output = AZ::VectorN::CreateZero(1024);
+        AZ::VectorN sourceVector = AZ::VectorN::CreateRandom(1024);
+        sourceVector *= 100.0f;
+        sourceVector -= 50.0f;
+        MachineLearning::Softmax(sourceVector, output);
+
+        // Sigmoid guarantees all outputs get squished between 0 and 1
+        for (AZStd::size_t iter = 0; iter < output.GetDimensionality(); ++iter)
+        {
+            ASSERT_GE(output.GetElement(iter), 0.0f);
+            ASSERT_LE(output.GetElement(iter), 1.0f);
+        }
+
+        // Additionally, the sum of all the elements should be <= 1, as softmax returns a probability distribution
+        const float totalSum = output.L1Norm();
+        // Between floating point precision and the estimates we use for exp(x), the total sum probability can be slightly greater than one
+        // We add a small epsilon to account for this error
```

Comment:
Should L1Norm include a clamp() to prevent the result from going over one?

Reply:
That's a good call, way better solution. I'll add that.

Reply:
https://github.com/o3de/o3de/pull/16643

Reply:
I don't know what I was thinking 🤣 It's absolutely wrong to clamp the output of L1Norm. We *actually* want to clamp the result of softmax. I'm going to abandon that L1 norm change as it breaks that function terribly.

Reply:
Also, if I somehow clamped SoftMax (by normalizing it again), that would honestly kind of break this particular test. The SoftMax should generate a probability distribution such that the sum of all probabilities equals to 1.

https://en.wikipedia.org/wiki/Softmax_function

We just want to make sure that the SoftMax math is actually correct by using this test.